keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,419 @@
+
+The device plugin support added in Kubernetes 1.8 makes it easy for vendors to dynamically export their resources through a plugin API without changing Kubernetes core code. Taking a further step, this document proposes a new resource abstraction API, ResourceClass, that can be used to describe, manage and consume such vendor specific and metadata rich resources in simple and portable ways.
+
+## Use Stories
+### As a cluster operator:
+- Nodes in my cluster has GPU HW from different generations. I want to classify GPU nodes into one of the three categories, silver, gold and platinum depending upon the launch timeline of the GPU family eg: Kepler K20, K80, Pascal P40, P100, Volta V100. I want to charge each of the three categories differently. I want to offer my clients 3 GPU rates/classes to choose from.<br/>
+**Motivation:** As time progresses in a cluster lifecycle, new advanced, high performance, expensive variants of GPUs gets added to the cluster nodes. At the same time older variants also co-exist. There are workloads which strictly wants latest GPUs and also there are workloads which are fine with older GPUs. But since there is a wide range of types, it will be hard to manage and confusing at the same time to have granularity at each GPU type. Grouping into few broad categories will be convenient to manage.<br/>
+**Can this be solved without resource classes:** A unique taint can be used to represent a category like silver. Nodes can be tainted accordingly depending upon the type of GPUs availability. User pods can use tolerations to steer workloads to the appropriate nodes. Now there are two problems. First, access control on tolerations and second, mechanism to quota resources. Though a new feature, [Pod Scheduling Policy][], is under design that can address first problem of access control on tolerations, there is no solution for second problem i.e quota control.  <br/>
+
+[Pod Scheduling Policy]: https://github.com/kubernetes/community/pull/1937
+
+**How Resource classes can solve this:** I, operator/admin, creates three resource classes: GPU-Platinum, GPU-Gold, GPU-Silver. Now since resource classes are quota controlled, end-user will be able to request resource classes only if quota is allocated.
+
+- I want a mechanism where it is possible to offer a group of devices, which are co-located on a single node and share a common property, as a single resource that can be requested in pod container spec. Example, N GPU units interconnected by NVLink or N cpu cores on same NUMA node.<br/>
```

Comment:
i think this section has mixed topology awareness with multi-type devices management, i suggest to separate
them to have focused discussion.

for the NUMA awareness,
should kubelet always try to schedule container(asking multiple cpus) to cpus on same NUMA node without user asking specifically? is there any penalty of doing so?

Reply:
@RobertKrawitz agree, for the example you are giving, the kubelet/cpumanager today does not guarantee placement decision, so it already has unpredictable performance in some sense.

this do brings up topology awareness scheduling requirement, i think it is in line with what i am suggesting to let kubelet handle NUMA scheduling natively, so it has consistent performance (some degree) for most cases (optimize for every case is not the goal and hard, affinity and anti-affinity on cpu level can help corner cases)